File: problems/leetcode/python/1099_two_sum_less_than_k.py
"""
Problem          : Two Sum Less Than K
Topics           : Sort, Search
Level            : Easy
URL              : https://leetcode.com/problems/two-sum-less-than-k
Description      : Given an array of integers, nums, and an integer k, find the maximum sum of two elements in nums
                    less than k. Otherwise, return −1 if no such pair exists.
Examples         :
                    Example 1:
                    Input: nums = [2,1,3,3,5], k = 4
                    Output: 3
                    Explanation: The maximum sum less than k is 3, i.e., 2+1 = 3.

                    Example 2:
                    Input: nums = [8,4,9,2,10,1], k = 9
                    Output: 6
                    Explanation: The maximum sum less than k is 6, i.e., 4+2 = 6.

                    Example 3:
                    Input: nums = [3,4,5,8,6,2], k = 5
                    Output: -1

                    Example 4:
                    Input: nums = [4,4,4,4,4,4,4,4], k = 12
                    Output: 8
"""


class Solution:
    def two_sum_less_than_k(self, nums: list[int], k: int) -> int:
        sorted(nums)
        if nums[0] >= k:
            return -1

        nums_len = len(nums)
        result = -1

        for i, num in enumerate(nums):
            j, want = -1, k - num
            left, right = i + 1, nums_len - 1
            while left <= right:
                mid = (left + right) // 2
                if nums[mid] < want:
                    j = mid
                    left = mid + 1
                else:
                    right = mid - 1

            if j > i:
                result = max(result, num + nums[j])

        return result


def run_tests():
    inputs = {
        "case-1": [
            [2, 1, 3, 3, 5],
            4,
        ],
        "case-2": [
            [8, 4, 9, 2, 10, 1],
            9,
        ],
        "case-3": [
            [3, 4, 5, 8, 6, 2],
            5,
        ],
        "case-4": [
            [4, 4, 4, 4, 4, 4, 4, 4],
            12,
        ],
    }
    outputs = {
        "case-1": 3,
        "case-2": 6,
        "case-3": -1,
        "case-4": 8,
    }

    solution = Solution()

    for case, input in inputs.items():
        result = solution.two_sum_less_than_k(input[0], input[1])
        assert result == outputs[case], f"{case}: expected {outputs[case]}, got {result}"

    print("✅ All tests passed!")


if __name__ == "__main__":
    run_tests()

File: problems/leetcode/go/1099_two_sum_less_than_k.go
package main

import (
	"fmt"
	"os"
	"sort"

	"github.com/ariefsibuea/algojo/libs/go/cmp"
)

func init() {
	register("TwoSumLessThanK", RunTestTwoSumLessThanK)
}

/**
 * Problem 			: Two Sum Less Than K
 * Topics           : Sort, Search
 * Level            : Easy
 * URL              : https://leetcode.com/problems/two-sum-less-than-k
 * Description      :
 * Examples         :
 * 					Example 1:
 * 					Input: nums = [2,1,3,3,5], k = 4
 * 					Output: 3
 * 					Explanation: The maximum sum less than k is 3, i.e., 2+1 = 3.
 *
 * 					Example 2:
 * 					Input: nums = [8,4,9,2,10,1], k = 9
 * 					Output: 6
 * 					Explanation: The maximum sum less than k is 6, i.e., 4+2 = 6.
 *
 * 					Example 3:
 * 					Input: nums = [3,4,5,8,6,2], k = 5
 * 					Output: -1
 *
 * 					Example 4:
 * 					Input: nums = [4,4,4,4,4,4,4,4], k = 12
 * 					Output: 8
 */

// twoSumLessThanK uses two pointers to solve the problem.
func twoSumLessThanK(nums []int, k int) int {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})

	if nums[0] >= k {
		return -1
	}

	result := -1
	left, right := 0, len(nums)-1

	for left < right {
		if nums[left]+nums[right] >= k {
			right -= 1
		} else {
			result = max(result, nums[left]+nums[right])
			left += 1
		}
	}

	return result
}

// twoSumLessThanKSolutionI uses binary search to solve the problem.
func twoSumLessThanKSolutionI(nums []int, k int) int {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})

	if nums[0] >= k {
		return -1
	}

	result := -1
	numsLen := len(nums)

	for i, num := range nums {
		j, t := -1, k-num
		left, right := i+1, numsLen-1
		for left <= right {
			mid := (left + right) / 2
			if nums[mid] < t {
				j = mid
				left = mid + 1
			} else {
				right = mid - 1
			}
		}

		if j > i {
			result = max(result, nums[i]+nums[j])
		}
	}

	return result
}

func RunTestTwoSumLessThanK() {
	testCases := map[string]struct {
		nums   []int
		k      int
		expect int
	}{
		"case-1": {
			nums:   []int{2, 1, 3, 3, 5},
			k:      4,
			expect: 3,
		},
		"case-2": {
			nums:   []int{8, 4, 9, 2, 10, 1},
			k:      9,
			expect: 6,
		},
		"case-3": {
			nums:   []int{3, 4, 5, 8, 6, 2},
			k:      5,
			expect: -1,
		},
		"case-4": {
			nums:   []int{4, 4, 4, 4, 4, 4, 4, 4},
			k:      12,
			expect: 8,
		},
	}

	for name, testCase := range testCases {
		fmt.Printf("RUN %s\n", name)
		result := twoSumLessThanK(testCase.nums, testCase.k)
		if !cmp.EqualNumbers(result, testCase.expect) {
			fmt.Printf("=== FAILED: expect = %v - got = %v\n", testCase.expect, result)
			os.Exit(1)
		}
		fmt.Printf("=== PASSED\n")
	}

	fmt.Printf("\n✅ All tests passed!\n")
}
